python/src/exactextract/writer.py

```diff
@@ -129,6 +129,169 @@ def features(self):
             return pd.DataFrame(self.fields, copy=False)
 
 
+class QGISWriter(Writer):
+    """Creates QGSVectorLayer"""
+
+    def __init__(self, *, srs_wkt=None):
+        super().__init__()
+        from qgis.core import QgsCoordinateReferenceSystem
+
+        self.fields = []
+        self.feats = []
+        self.geom_type = "unknown"
+        self.fields_set = False
+
+        self.crs = QgsCoordinateReferenceSystem()
+        self.crs.createFromWkt(srs_wkt)
+
+    def add_operation(self, op):
+        self.fields.append(op.name)
+
+    def add_column(self, col_name):
+        self.fields.append(col_name)
+
+    def write(self, feature):
+        import json
+
+        from qgis.core import QgsFeature, QgsJsonUtils
+
+        f = JSONFeature()
+        feature.copy_to(f)
+
+        if not self.fields_set:
+            # test every property vs fields type and set correct type
+            self.fields = self._set_fields_types(self.fields, f)
+            self.fields = self._convert_to_QgsFields(self.fields)
+            self.geom_type = self._get_geometry_type(f)
+            self.fields_set = True
+
+        # create a QgsFeature and set fields
+        fet = QgsFeature()
+        fet.setFields(self.fields)
+        # populate QgsFeature with values
+        for field_name, value in f.feature["properties"].items():
+            field_type = self.fields.at(
+                self.fields.indexFromName(field_name)
+            ).typeName()
+            if field_type == "Array":
+                value = str(value)
+            fet.setAttribute(field_name, value)
+        # set id if present
+        if "id" in f.feature:
+            fet.setAttribute("id", f.feature["id"])
+        # set geometry based on JSON geometry
+        fet.setGeometry(
+            QgsJsonUtils.geometryFromGeoJson(json.dumps(f.feature["geometry"]))
```

Comment:
I'm getting `AttributeError: type object 'QgsJsonUtils' has no attribute 'geometryFromGeoJson'` with `qgis.core.Qgis.QGIS_VERSION` = '3.34.3-Prizren'

Better to go through WKB if that's possible, I think.

Reply:
I will try to go with `QGSGeometry.fromWkb`.

Reply:
Inside `write` method `feature` is of type `JSONFeature`. Do you know how I can approach the conversion of JSON geometry into WKB without using `shapely`?

Reply:
See how `GDALFeature` uses the `set_geometry_format` method to ask for geometry in WKB format, and `set_geometry` to handle it.

https://github.com/isciences/exactextract/blob/master/python/src/exactextract/feature.py#L100

Reply:
Also, I know this API is completely undocumented, so nice job figuring this all out!

Reply:
I'm trying to wrap my head around this. Can you point me to where the `set` and `set_geometry` are called? If it's done at the C++ level I will probably need some help to understand the proper way and place to create `QgsFeature` in `QGISFeature`.

Reply:
This is the current state of code that I came up with. It's not working now as the `set` and `set_geometry` seem to be never executed. Moreover, I don't understand the type of feature that comes into `QGISFeature` when used in 'QGISWriter`.
writer.py: https://gist.github.com/JakubCha/bb7b04d0062e97d1231384a74dc241bb
feature.py: https://gist.github.com/JakubCha/ae31cf6865e3b8e34817d539f07b21f1
The code that's there right now doesn't make sense, probably.

Reply:
Thanks! I'll try to check this out in the next day or two.

Reply:
Here's a candidate: https://github.com/dbaston/exactextract/commit/af6ed5e11b4e7c0777e84e44ff4ecbc156b2a627

Reply:
So it was all about `copy_to` 👍 
I made a new commit below which fixes array results.